crates/services/sync/src/import.rs

```diff
@@ -182,83 +228,96 @@ where
+                let shutdown_guard = shutdown_guard.clone();
+                let shutdown_signal = shutdown_signal.clone();
+                tokio::spawn(async move {
+                    // Hold a shutdown sender for the lifetime of the spawned task
+                    let _shutdown_guard = shutdown_guard.clone();
+                    let mut shutdown_signal = shutdown_signal.clone();
+                    tokio::select! {
+                        // Stream a batch of blocks
+                        blocks = stream_block_batch => blocks,
+                        // If a shutdown signal is received during the stream, terminate early and
+                        // return an empty response
+                        _ = shutdown_signal.while_started() => (vec![], None)
+                    }
+                }).then(|task| async { task.map_err(ImportError::JoinError) })
+            })
+            // Request up to `block_stream_buffer_size` transactions from the network.
+            .buffered(params.block_stream_buffer_size)
+            // Continue the stream unless an error occurs.
+            // Note the error will be returned but the stream will close.
+            .into_scan_err()
+            .scan_err()
+            // Continue the stream until the shutdown signal is received.
+            .take_until({
+                let mut s = shutdown.clone();
                 async move {
-                    let (peer_id, block) = res?;
-
-                    let res = execute_and_commit(executor.as_ref(), &state, block).await;
+                    let _ = s.while_started().await;
+                    tracing::info!("In progress import stream shutting down");
+                }
+            })
+            // Then execute and commit the block
+            .zip(context_generator)
+            .then(
+                |(blocks_result, (peer, state, p2p, executor))| async move {
+                    let (sealed_blocks, error) = blocks_result?;
+                    let sealed_blocks = futures::stream::iter(sealed_blocks);
+                    let res = sealed_blocks.then(|sealed_block| async {
+                        execute_and_commit(executor.as_ref(), &state, sealed_block).await
+                    }).try_collect::<Vec<_>>().await.and_then(|v| error.map_or(Ok(v), Err));
                     match &res {
                         Ok(_) => {
-                            let _ = p2p.report_peer(peer_id.clone(), PeerReportReason::SuccessfulBlockImport)
-                                .await
-                                .map_err(|e| tracing::error!("Failed to report successful block import for peer {:?}: {:?}", peer_id, e));
+                            report_peer(p2p.as_ref(), peer.clone(), PeerReportReason::SuccessfulBlockImport).await;
                         },
                         Err(e) => {
                             // If this fails, then it means that consensus has approved a block that is invalid.
```

Comment:
If I recall correctly, there was a reason that we didn't report the peer here, but I don't have all the context. @MitchTurner - should we report the peer in this case?

Reply:
I guess because, in this case, it means that the network doesn't work anymore=) But if we don't decrease the reputation, it means we infinitely will jibe blocks between peers. When decreasing, it will disconnect them.